from collections import namedtuple
import os

CampareeConstants = namedtuple('Constants',
                               ['CAMPAREE_ROOT_DIR',
                                'CAMPAREE_OUTPUT_DIR_NAME',
                                'DEFAULT_STAR_OUTPUT_PREFIX',
                                'DEFAULT_STAR_BAM_FILENAME',
                                'INTRON_OUTPUT_FILENAME',
                                'INTRON_OUTPUT_ANTISENSE_FILENAME',
                                'INTERGENIC_OUTPUT_FILENAME',
                                'VARIANTS_FINDER_OUTPUT_FILENAME',
                                'VARIANTS_FINDER_LOG_FILENAME',
                                'VARIANTS_COMPILATION_OUTPUT_FILENAME',
                                'VARIANTS_COMPILATION_LOG_FILENAME',
                                'BEAGLE_OUTPUT_PREFIX',
                                'BEAGLE_OUTPUT_FILENAME',
                                'BEAGLE_LOG_FILENAME',
                                'GENOMEBUILDER_OUTPUT_PREFIX',
                                'GENOMEBUILDER_SEQUENCE_FILENAME_PATTERN',
                                'GENOMEBUILDER_INDEL_FILENAME_PATTERN',
                                'GENOMEBUILDER_MAPPER_FILENAME_PATTERN',
                                'GENOMEBUILDER_LOG_FILENAME',
                                'UPDATEANNOT_OUTPUT_FILENAME_PATTERN',
                                'UPDATEANNOT_LOG_FILENAME_PATTERN',
                                'TRANSCRIPTOMES_LOG_FILENAME'])

CampareeConstants.__doc__ = """
Provides a list of constants specific to the CAMPAREE pipeline.
These include input/output/logging filenames used by various
steps in the pipeline.
"""

_CAMPAREE_ROOT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
_DEFAULT_STAR_OUTPUT_PREFIX="genome_alignment."
CAMPAREE_CONSTANTS = \
    CampareeConstants(CAMPAREE_ROOT_DIR=_CAMPAREE_ROOT_DIR,
                      # Name of sub-directory containing CAMPAREE results in output directory.
                      CAMPAREE_OUTPUT_DIR_NAME="CAMPAREE",
                      # Prefix this pipeline assigns to the STAR output files (excluding path),
                      # if the user does not provide pre-aligned BAM files. This is part of what
                      # is passed STAR as a command line parameter.
                      DEFAULT_STAR_OUTPUT_PREFIX=_DEFAULT_STAR_OUTPUT_PREFIX,
                      # Full name (exlcuding path) of STAR-created BAM file generated by this
                      # pipeline. The same filename is used across all samples.
                      DEFAULT_STAR_BAM_FILENAME=f"{_DEFAULT_STAR_OUTPUT_PREFIX}Aligned.sortedByCoord.out.bam",
                      INTRON_OUTPUT_FILENAME="intron_quantifications.txt",
                      INTRON_OUTPUT_ANTISENSE_FILENAME="intron_antisense_quantifications.txt",
                      INTERGENIC_OUTPUT_FILENAME="intergenic_quantifications.txt",
                      # Name of file where VariantsFinderStep output is stored.
                      VARIANTS_FINDER_OUTPUT_FILENAME="variants.txt",
                      # Name of file where VariantsFinderStep logging is stored.
                      VARIANTS_FINDER_LOG_FILENAME="VariantsFinderStep.log",
                      # Name of file where VariantsCompilationStep output is stored.
                      VARIANTS_COMPILATION_OUTPUT_FILENAME="all_variants.vcf",
                      # Name of file where VariantsCompilationStep logging is stored.
                      VARIANTS_COMPILATION_LOG_FILENAME="VariantsCompilationStep.log",
                      # Prefix assigned to all beagle output files. This is part of what is passed
                      # to beagle as a command line parameter.
                      BEAGLE_OUTPUT_PREFIX="beagle",
                      # Name of output file generated by the beagle program.
                      BEAGLE_OUTPUT_FILENAME="beagle.vcf.gz",
                      # Name of file where BeagleStep logging is stored.
                      BEAGLE_LOG_FILENAME="BeagleStep.log",
                      #Prefix for names of all (non-log) files generated by GenomeBuilderStep
                      GENOMEBUILDER_OUTPUT_PREFIX="custom_genome",
                      #String pattern to construct sequence FASTA filename generated by GenomeBuilderStep.
                      GENOMEBUILDER_SEQUENCE_FILENAME_PATTERN='{genome_output_file_stem}_{genome_name}.fa',
                      #String pattern to construct indel filename generated by GenomeBuilderStep
                      GENOMEBUILDER_INDEL_FILENAME_PATTERN='{genome_output_file_stem}_indels_{genome_name}.txt',
                      #String pattern to construct reference mapping filename generated by GenomeBuilderStep
                      GENOMEBUILDER_MAPPER_FILENAME_PATTERN='{genome_output_file_stem}_to_parental_coordinate_mapping_{genome_name}.txt',
                      #Name of file where GenomeBuilderStep logging is stored
                      GENOMEBUILDER_LOG_FILENAME="GenomeBuilderStep.log",
                      #Name of updated annotation file generated UpdateAnnotationForGenomeStep.
                      UPDATEANNOT_OUTPUT_FILENAME_PATTERN='updated_annotation_{genome_name}.txt',
                      #Name of file where UpdateAnnotationForGenomeStep logging is stored
                      UPDATEANNOT_LOG_FILENAME_PATTERN='UpdateAnnotationForGenomeStep_{genome_name}.log',
                      TRANSCRIPTOMES_LOG_FILENAME="TranscriptQuantificatAndMoleculeGenerationStep.log"
                     )

CAMPAREE_VERSION="0.1.0"
